File: server/controllers/authController.js
const { db, auth } = require("../firebase/firebase");
const register = async (req, res) => {
  const { email, password, username } = req.body;

  try {
    const usernameValidation = await db
      .collection("users")
      .where("username", "==", username)
      .get();

    if (!usernameValidation.empty) {
      return res.status(400).json({ error: "This username is taken." });
    }

    const emailValidation = await auth.getUserByEmail(email).catch(() => null);

    if (emailValidation) {
      return res
        .status(400)
        .json({ error: "An account with this email already exists." });
    }

    const user = await auth.createUser({
      email,
      password,
      displayName: username,
    });

    await db.collection("users").doc(user.uid).set({
      username,
      email,
      createdAt: new Date().toISOString(),
    });

    return res.status(201).json({
      message: "User registered successfully!",
    });
  } catch (error) {
    return res.status(400).json({ error: error.message });
  }
};

module.exports = {
  register,
};

File: server2/controllers/auth_controller.py
from flask import request, jsonify
from firebase.firebase_config import db, firebase_auth
from datetime import datetime


def register():

    data = request.json

    email = data.get("email")
    password = data.get("password")
    username = data.get("username")

    try:

        # Username validation
        username_validation = (
            db.collection("users")
            .where("username", "==", username)
            .get()
        )

        if len(username_validation) > 0:
            return jsonify({
                "error": "This username is taken."
            }), 400

        # Email validation
        try:
            firebase_auth.get_user_by_email(email)

            return jsonify({
                "error": "An account with this email already exists."
            }), 400

        except:
            pass

        # Create Firebase user
        user = firebase_auth.create_user(
            email=email,
            password=password,
            display_name=username
        )

        # Save user in Firestore
        db.collection("users").document(user.uid).set({
            "username": username,
            "email": email,
            "createdAt": datetime.utcnow().isoformat()
        })

        return jsonify({
            "message": "User registered successfully!"
        }), 201

    except Exception as error:
        return jsonify({
            "error": str(error)
        }), 400